## Permutations, Combinations, Conditional Probability and Partitioning Complex Events



In [6]:
#permutations - goin

## What's the difference between a permutation and a combination?

## How many permutations are there for a standard padlock?
> Hint: (there are 40 numbers on a padlock and use 3 numbers.)

In [7]:
import pandas as pd

## How many unique 3 topping pizzas can you make from the following ingredients:
    * Mushrooms
    * Pepperoni
    * Onion
    * Peppers
    * Ham
    * Pineapple
    * Sausage
    * Olives
    
> Fun question: which bis the worst?

In [8]:
import math
# 8 ingredients
# don't want to repeat
math.factorial(8)/(factorial(5)*factorial(3))

NameError: name 'factorial' is not defined

## Explaining the Intuition Behind Combinations


How would you compute the number of 3-topping combinations from the ingredients above by hand? Explain the rational behind your computations.

## Conditional Probability

### When do we compute conditional probabilities? 

### Mushroom dataset

To discuss conditional probability, let's look at a modified version of the Mushroom dataset from UCI [here](https://archive.ics.uci.edu/ml/datasets/Mushroom). Each row in this dataset corresponds to one observation (one mushroom). 

The modified dataset includes 4 variables:

* **edible-poisonous**
    * This categorical variable can have one of two values: if the mushroom is edible, "edible". If not, "poisonous"

* **bruised**
    * This is a Boolean variable that can assume either one of two values, True or False.

* **gill-spacing**
    * This categorical variable can have one of three values: "close", "crowded", or "distant"
    
* **stalk-shape**
    * This categorical variable can have one of two values: "enlarging" or "tapering"
* **stalk-color-above-ring**
    * This categorical variable can have one of nine values:  "brown","buff","cinnamon","gray","orange", "pink","red","white" or "yellow"

* **stalk-color-below-ring**
    * This categorical variable can have one of nine values:  "brown","buff","cinnamon","gray","orange", "pink","red","white" or "yellow"

* **gill-color**
    * This categorical variable can have one of twelve values: "black","brown","buff","chocolate","gray", "green","orange","pink","purple","red", "white" or "yellow" 



In [9]:
import pandas as pd

df = pd.read_csv('Mushrooms_cleaned.csv')
df.head()

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
0        poisonous        close   enlarging                  white   
1           edible        close   enlarging                  white   
2           edible        close   enlarging                  white   
3        poisonous        close   enlarging                  white   
4           edible      crowded    tapering                  white   

  stalk-color-below-ring gill-color  bruised  
0                  white      black     True  
1                  white      black     True  
2                  white      brown     True  
3                  white      brown     True  
4                  white      black    False

In [28]:
df

edible-poisonous gill-spacing stalk-shape stalk-color-above-ring  \
0           poisonous        close   enlarging                  white   
1              edible        close   enlarging                  white   
2              edible        close   enlarging                  white   
3           poisonous        close   enlarging                  white   
4              edible      crowded    tapering                  white   
5              edible        close   enlarging                  white   
6              edible        close   enlarging                  white   
7              edible        close   enlarging                  white   
8           poisonous        close   enlarging                  white   
9              edible        close   enlarging                  white   
10             edible        close   enlarging                  white   
11             edible        close   enlarging                  white   
12             edible        close   enlarging                  white   
13          poisonous        close   enlarging                  white   
14             edible      crowded    tapering                  white   
15             edible        close   enlarging                  white   
16             edible      crowded    tapering                  white   
17          poisonous        close   enlarging                  white   
18          poisonous        close   enlarging                  white   
19          poisonous        close   enlarging                  white   
20             edible        close   enlarging                  white   
21          poisonous        close   enlarging                  white   
22             edible        close   enlarging                  white   
23             edible        close   enlarging                  white   
24             edible        close   enlarging                  white   
25          poisonous        close   enlarging                  white   
26             edible        close   enlarging                  white   
27             edible        close   enlarging                  white   
28             edible        close   enlarging                  white   
29             edible      crowded    tapering                  white   
...               ...          ...         ...                    ...   
8094           edible      crowded   enlarging                  white   
8095        poisonous        close   enlarging               cinnamon   
8096           edible      crowded   enlarging                  white   
8097        poisonous        close    tapering                   pink   
8098        poisonous        close    tapering                  white   
8099           edible      crowded   enlarging                  white   
8100           edible        close   enlarging                 orange   
8101        poisonous        close    tapering                   pink   
8102           edible        close   enlarging                 orange   
8103           edible        close   enlarging                 orange   
8104           edible        close   enlarging                 orange   
8105           edible        close   enlarging                 orange   
8106           edible        close   enlarging                 orange   
8107           edible        close   enlarging                 orange   
8108        poisonous        close    tapering                   pink   
8109           edible      crowded   enlarging                  white   
8110           edible        close   enlarging                 orange   
8111           edible      crowded   enlarging                  white   
8112           edible        close   enlarging                 orange   
8113        poisonous        close    tapering                   pink   
8114        poisonous        close   enlarging               cinnamon   
8115           edible        close   enlarging                 orange   
8116        poisonous        close    tapering                 

#### If you picked a row from this dataset at random, what is the probability it corresponds to a bruised mushroom? 

In [10]:
# df.shape
# (df['bruised'] == True).sum()/8124

df[df['bruised'] == True].shape[0]

3376

#### What is the probability you pick a row corresponding to a mushroom that is bruised _AND_ edible? 

In [11]:
# edible = df[df['edible-poisonous'] == 'edible']

# (edible['bruised'] == True).sum()/8124

df[(df['bruised'] == True) & (df['edible-poisonous'] == 'edible')].shape[0]/df.shape[0]


0.33874938453963566

#### What is the probability of picking an edible mushroom given it is bruised, $P(edible|bruised)$? 

In [12]:
bruised_mushrooms = df[df['bruised'] == True]
bruised_mushrooms.shape
(bruised_mushrooms['edible-poisonous'] == 'edible').sum()/bruised_mushrooms.shape[0]

0.8151658767772512

#### What is the probability of picking a bruised mushroom given it is edible, $P(\text{bruised | edible})$? 

In [13]:
edible_mushrooms = df[df['edible-poisonous'] == 'edible']
edible_mushrooms.shape #4208

(edible_mushrooms['bruised'] == True).sum()/edible_mushrooms.shape[0]


0.6539923954372624

In [41]:
df2 = (df[(df['stalk-color-above-ring'] == 'orange')|
         (df['stalk-color-below-ring'] == 'orange')| 
         (df['gill-color'] == 'orange')])

df3 = (df2['edible-poisonous'] == 'edible')
df3.shape

(192,)

In [23]:
df['gill-color']

0        black
1        black
2        brown
3        brown
4        black
5        brown
6         gray
7        brown
8         pink
9         gray
10        gray
11       brown
12       white
13       black
14       brown
15       black
16       black
17       brown
18       brown
19       black
20       black
21       brown
22       black
23       white
24        gray
25       brown
26       brown
27       white
28       black
29       brown
         ...  
8094      gray
8095    yellow
8096     white
8097      buff
8098      buff
8099     white
8100    orange
8101      buff
8102    yellow
8103    yellow
8104    yellow
8105    yellow
8106    orange
8107    yellow
8108      buff
8109     white
8110    orange
8111      pink
8112    orange
8113      buff
8114    yellow
8115    yellow
8116      buff
8117      buff
8118      buff
8119    yellow
8120    yellow
8121     brown
8122      buff
8123    yellow
Name: gill-color, Length: 8124, dtype: object

### Intuition behind conditional probability: 

How do you compute the probability that mushrooms are edible given they are bruised? 

When you ask the question "what is the probability that the mushrooms are edible and bruised?", the sample space originally contains all 8124 rows of mushrooms. 

<img src="images/Image_72_Cond4.png" width="300">

However, to compute the probability that the mushrooms are edible given they are bruised, you need to consider the reduced size of the sample space. 

In the image above, S is the universe of all mushrooms in the dataset, A is the set of mushrooms that are edible, and B is the set of mushrooms that are bruised.

* When you ask the question "what is the probability that the mushrooms are edible given the mushrooms are bruised?", you have effectively reduced the size of the sample space to include only those mushrooms that are bruised. 

* Given that mushrooms are bruised, the only way for the mushrooms to be edible is for these mushrooms to fall in the intersection of the set of mushrooms that are edible _and_ the set of mushrooms that are bruised , $P(edible \cap bruised)$.  

* To account for the smaller sample space, you divide the probability mushrooms are edible and bruised by the probability the mushrooms are bruised: $$\large P(edible|bruised) = \frac{P(edible \cap bruised)}{P(bruised)}$$




## Partitioning Complex Events

You're not really a mushroom expert, but you can see a bunch of orange spots all over the mushroom in your hand. Given the data at your disposal, what's the probability that the mushroom is edible?



## Summary


In this lesson, you reviewed 4 major foundational concepts for probability: permutations, combinations, conditional probability and partitioning complex events. Remember that your standard padlock should be more accurately called a permutation lock! Order matters for permutations, whereas only the members of the set are important for combinations. Conditional probability investigates the odds of an event occurring given other information. In these instances, the universal set of possibilities reflects the given information. In the mushroom example, the probability of a mushroom being edible given that it is bruised can be computed by dividing the probability that it a mushroom is both edible AND bruised, by the probability that it is bruised. Mathematically: $$\large P(edible|bruised) = \frac{P(edible \cap bruised)}{P(bruised)}$$. Finally, you investigated partitioning complex events. Often, complex events can be broken into constituent parts, and the total probability can be calculated by combining these smaller events.

## Additional Resources

## Challenge Problem

Let's take some time and review questions like those from the [dsc-law-of-total-probability-lab](https://github.com/learn-co-curriculum/dsc-law-of-total-probability-lab).  

According to the CDC, [14% of Americans currently smoke, 15.8% of males and 12.2% of females](https://www.cdc.gov/tobacco/data_statistics/fact_sheets/adult_data/cig_smoking/index.htm). According the the American Lung Association, [men who smoke are 23 times more likely to smoke then never-smokers, and women are 13 times as likely](https://www.lung.org/lung-health-and-diseases/lung-disease-lookup/lung-cancer/resource-library/lung-cancer-fact-sheet.html). The American Cancer Society estimates that [the lifetime risk of developing lung cancer is 6.85% for males and 5.95% for females](https://www.cancer.org/cancer/cancer-basics/lifetime-probability-of-developing-or-dying-from-cancer.html). Currently, the census estimates that [women are 50.8% of the population](https://www.census.gov/quickfacts/fact/table/US/PST045218). 

What is the risk of lung cancer for non-smokers? Non-smoker males? Non-smoker females?

> To learn more about lung-cancer risks for non-smokers, see https://www.cancer.org/latest-news/why-lung-cancer-strikes-nonsmokers.html.

## Additional Conditional Probability Practice

What's the probability that a mushroom is poisonous if it has close gill spacing and a tapering stalk?

$$\large P(edible|close \cap tapering) = \frac{P(edible \cap close \cap tapering)}{P(close \cap tapering)}$$